libcoap/src/types.rs

```diff
@@ -258,153 +270,545 @@ impl From<coap_uri_scheme_t> for CoapUriScheme {
+    /// assert_eq!(uri.port(), Some(4711));
+    /// assert_eq!(uri.path(), Some("foo/bar".as_bytes()));
+    /// assert_eq!(uri.query(), Some("answer=42".as_bytes()));
+    /// assert!(uri.is_proxy());
+    ///
+    /// # Result::<(), UriParsingError>::Ok(())
+    /// ```
+    pub fn try_from_str_proxy(uri_str: &str) -> Result<CoapUri, UriParsingError> {
+        // SAFETY: coap_split_proxy_uri is one of the allowed functions.
+        unsafe { CoapUri::create_parsed_uri(CString::new(uri_str)?, coap_split_proxy_uri, true) }
     }
 
     /// Attempts to convert a [Url] into a [CoapUri].
     ///
     /// # Errors
-    /// May fail if the provided Url has an invalid scheme.
-    pub fn try_from_url(url: Url) -> Result<CoapUri, UriParsingError> {
-        let path: Vec<String> = url
-            .path()
-            .split('/')
-            .map(String::from)
-            .filter(|v| !v.is_empty())
-            .collect();
-        let path = if path.is_empty() { None } else { Some(path) };
-
-        let query: Vec<String> = url.query_pairs().map(|(k, v)| format!("{}={}", k, v)).collect();
-        let query = if query.is_empty() { None } else { Some(query) };
-        Ok(CoapUri {
-            scheme: Some(CoapUriScheme::from_str(url.scheme())?),
-            host: url.host().map(|h| h.into()),
-            port: url.port(),
-            path,
-            query,
-        })
+    /// May fail if the provided Url is not a valid URI supported by libcoap or if the URI
+    /// components exceed maximum lengths (see the struct level documentation).
+    #[cfg(feature = "url")]
+    pub fn try_from_url(url: &Url) -> Result<CoapUri, UriParsingError> {
+        Self::try_from_str(url.as_str())
+    }
+
+    /// Attempts to convert a [Url] into a proxy [CoapUri].
+    ///
+    /// # Errors
+    /// May fail if the provided Url is not a valid proxy URI supported by libcoap or if the URI
+    /// components exceed maximum lengths (see the struct level documentation).
+    #[cfg(feature = "url")]
+    pub fn try_from_url_proxy(url: &Url) -> Result<CoapUri, UriParsingError> {
+        Self::try_from_str_proxy(url.as_str())
     }
 
     /// Returns the scheme part of this URI.
-    pub fn scheme(&self) -> Option<&CoapUriScheme> {
-        self.scheme.as_ref()
+    pub fn scheme(&self) -> Option<CoapUriScheme> {
+        // URIs can either be absolute or relative. If they are relative, the scheme is also not
+        // set (but defaults to CoAP as the default enum value is 0).
+        self.host()?;
+        Some(CoapUriScheme::from_raw_scheme(self.raw_uri.scheme))
     }
 
     /// Returns the host part of this URI.
-    pub fn host(&self) -> Option<&CoapUriHost> {
-        self.host.as_ref()
+    pub fn host(&self) -> Option<&[u8]> {
+        let raw_str = self.raw_uri.host;
+        if raw_str.length == 0 {
+            return None;
+        }
+        Some(unsafe { std::slice::from_raw_parts(raw_str.s, raw_str.length) })
```

Comment:
Should this have a `SAFETY` comment? (Also applies to other instances in this file.)